WIP: Add more features to nordic IO

calum-chamberlain opened the pull request:
# Work in progress, do not merge yet!
### What does this PR do?

This PR attempts to add more functionality to Nordic format IO operations. This will add in:
~~- [x] Read *hyp* error estimates and convert to confidenceEllipsoid;~~
~~- [x] Write confidenceEllipsoid to *hyp* uncertainty estimates;~~
- [x] Read focal mechanism solutions;
- [x] Write focal mechanism solutions;
- [x] Read moment tensor solutions;
- [x] Write moment tensor solutions.
- [x] latin1 encoded reading (currently raises a `UnicodeDecodeError` in Py3 and other errors in py2) - how should format be detected, `chardet`, or should the user specify if there is an error? Could ask Jens what formats are accepted and try then?
~~-  [ ] More `xyz_to_confidence_ellipsoid` and `confidence_ellipsoid_to_xyz` to `obspy.core.event.origin`.~~
- [x] Review comments.

### Why was it initiated?  Any relevant Issues?

The ability to read *Hyp* derived errors was requested by @shicks-seismo in issue #1909 

### PR Checklist
- [x] This PR is not directly related to an existing issue (which has no PR yet).
- [ ] All tests still pass.
- [x] Any new features or fixed regressions are be covered via new tests.
- [x] Any new or changed features have are fully documented.
- [x] Significant changes have been added to `CHANGELOG.txt` .
~~- [ ] First time contributors have added your name to `CONTRIBUTORS.txt` .~~

calum-chamberlain commented:
This PR is more to document progress, and ask for help on converting between covariance matrix and confidenceEllipsoid (and back).  I have made a stab at this in [`obspy.io.nordic.core.xyz_to_confidence_ellipsoid`](https://github.com/obspy/obspy/pull/1924/files#diff-4bc6ea61b90645c5ad3a04a2f843d8c8R249) and the sister function: [`obspy.io.nordic.core.confidence_ellipsoid_to_xyz`](https://github.com/obspy/obspy/pull/1924/files#diff-4bc6ea61b90645c5ad3a04a2f843d8c8R293), but I'm failing at the moment - welcome any changes here!

I don't think this PR is urgent.

calum-chamberlain commented:
Reading now implemented, all tests pass expect `confidenceEllipsoid` conversion, which I will work to fix at some point in the next couple of weeks.

calum-chamberlain commented:
@shicks-seismo, I think the reading in is working, do you have any expected results for confidence ellipsoids from x, y, z and covariance? I'm still stuck on writing (going back the other way), I can get the right numbers but not the right order.  Any help appreciated.

megies commented:
> This PR is more to document progress, and ask for help on converting between covariance matrix and confidenceEllipsoid (and back). I have made a stab at this in obspy.io.nordic.core.xyz_to_confidence_ellipsoid and the sister function: obspy.io.nordic.core.confidence_ellipsoid_to_xyz, but I'm failing at the moment - welcome any changes here!

I think these helper functions should eventually be moved to `obspy.core.event.origin` for general use. No hurry though, can also be refactored at the end when everything is ready, and we won't do 1.2.0 until march/april.

shicks-seismo commented:
Hi @calum-chamberlain, sorry for the late reply. I am not really using the error ellipsoids so I don't really have any expectations here.
Note my new PR @ #1991 regarding longitude and depth reading.

calum-chamberlain commented:
Brief update on this: I have been in touch with Jens Havskov and he thinks they (the SEISAN team) have conversions from x, y, z to ellipse (oid?), but not the reverse, which is still the part I am struggling with. He also tried out a nordic file with some interesting things that I didn't account for, but have included here.

One other thing his file highlighted is that this module doesn't currently account for different encodings - his file was encoded as latin1, which breaks these functions.  Do we have a good/standard way of checking encoding ([`chardet`](https://chardet.github.io/) seems like the tool, but would slow things down most of the time)? As a prefred way of doing it, I have asked Jens what the standard encodings are and then I can add in a `try: .. except:` clause to try different encodings. Thoughts @krischer @megies ?

megies commented:
> One other thing his file highlighted is that this module doesn't currently account for different encodings - his file was encoded as latin1, which breaks these functions. Do we have a good/standard way of checking encoding (chardet seems like the tool, but would slow things down most of the time)? As a prefred way of doing it, I have asked Jens what the standard encodings are and then I can add in a try: .. except: clause to try different encodings. Thoughts @krischer @megies ?

without looking at the details.. if there really are different encodings used in the wild, the easiest option would be to at least add an `encoding` kwarg that can be specified by the user (maybe defaulting to `ASCII`), so at least users can read such files without hacking the code

krischer commented:
I think trying a few default encodings would be good and simple first step. Otherwise you could raise an error and tell users that they have to specify the encoding and maybe even point them to the `chardet` library to figure out which encoding they have.

calum-chamberlain commented:
The word from Jens is that nordic files in seisan are read in as bytes and decoded as single-byte characters, which seems to be consistent with everything being `latin-1` encoded.  This appears to only have been an issue for some of the less-used characters (like: Ø, in the included `dos-file.sfile` test file).

Reading everything using:
```python
f = open(sfile, 'r', encoding='latin-1')
```
seems to work fine and pass all my tests...

krischer commented:
Okay good. I still think an optional `encoding` argument would be useful for example when somebody manually edits a file and accidentally saves it in a different encoding. Also its really trivial to implement.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -57,19 +57,22 @@ def __init__(self, value):
     :type sfile: str
     :param sfile: Path to sfile
+    :type encoding: str
+    :param encoding: Encoding for file, used to decode from bytes to string
```
This sounds a bit too technical. How about just "Encoding of the file"?

krischer commented on CHANGELOG.txt:
```diff
@@ -151,6 +151,11 @@ master:
+ - obspy.io.nordic:
+    * Add ability to read and write focal mechanism, hypocentral error, moment
+      tensor information
+    * Add explicit warnings about lines not yet able to read.
```
I don't think that's a complete sentence.

krischer commented on obspy/io/nordic/tests/test_nordic.py:
```diff
@@ -412,6 +432,8 @@ def test_write_select(self):
                 warnings.simplefilter('ignore', UserWarning)
                 write_select(cat, filename=tf.name)
+            if not _is_sfile(tf.name):
+                raise(NordicParsingError("format is wrong somehow"))
```
Replace with `self.assertTrue(_is_sfile(tf.name))`.

krischer commented on obspy/io/nordic/tests/test_nordic.py:
```diff
@@ -546,6 +587,18 @@ def test_distance_conversion(self):
+        errors = {'x_err': 1., 'y_err': 1.5, 'z_err': 2.0,
+                  'xy_cov': 0.0, 'xz_cov': 0.0, 'yz_cov': 0.0}
+        confidence_ellipsoid = xyz_to_confidence_ellipsoid(errors)
+        print(confidence_ellipsoid)
```
Please remove the print statement.

calum-chamberlain replied:
Will do once its working!

krischer commented on obspy/io/nordic/tests/test_nordic.py:
```diff
@@ -546,6 +587,18 @@ def test_distance_conversion(self):
+        print(confidence_ellipsoid)
+        errors_back = confidence_ellipsoid_to_xyz(confidence_ellipsoid)
+        self.assertEqual(errors, errors_back)
+        # TODO: Test mapping for known result.
```
This should really be done as its otherwise hard to really trust the calculations - so the test just shows that one function is the inverse of the other which is necessary but not sufficient IMHO. For simple cases it should also be fairly easy to do.

calum-chamberlain replied:
Agree - sorry, this still isn't ready - the confidence ellipsoid stuff still doesn't get the right result for the reverse (confidence_ellipsoid_to_xyz), so I haven't written more tests yet.

krischer replied:
Yea I should have read the first post 🙈 Sorry about that!

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -28,19 +28,26 @@
 
+accepted_tags = ['1', '6', '7', 'E', ' ', 'F', 'M', '3']
+# List of currently implemented line-endings, which in Nordic mark what format
+# info in that line will be.
```
Should this go above the code?

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -50,51 +57,74 @@ def __init__(self, value):
+def _get_line_tags(f):
+    f.seek(0)
+    line = f.readline()
+    if len(line.rstrip('\n').rstrip('\r')) != 80:
```
Just `.rstrip('\n\r')` should be enough. Stripping always removes as many of those. Just using `.rstrip()` should actually also be enough as it by default will remove any ASCII whitespace I believe.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -50,51 +57,74 @@ def __init__(self, value):
+        elif line_id in accepted_tags:
+            tags.update({line_id: [(line, i)]})
+        else:
+            UserWarning("Lines of type %s have not been implemented yet, "
```
You have to use `warnings.warn()` to actually raise a warning.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -50,51 +57,74 @@ def __init__(self, value):
+        if line_id in tags.keys():
+            tags[line_id].append((line, i))
+        elif line_id in accepted_tags:
+            tags.update({line_id: [(line, i)]})
```
This would be about the same number of lines but it feels a bit more elegant to me:

```python
tags = collections.defaultdict(list)
...
if line_id in accepted_tags:
    tags[line_id].append((line, i))
```

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -322,40 +414,113 @@ def _readheader(f):
+    if new_event.origins[0].depth:
+        new_event.origins[0].depth *= 1000
+    # new_event.depth_ind = line[44]
+    # new_event.loc_ind = line[45]
```
Why are those commented?

calum-chamberlain replied:
Commented because I don't know where to put them in the Event format, any ideas? They are single character identifies that the user defines.
depth_ind is usually (AFAIK) one of `(' ', 'F', 'S')` where `'F'` marks fixed depth, `'S'` marks that it is a starting value (for inversion I presume).
loc_ind looks like it is either `(' ', '*')` where `*` means "do not locate".

krischer replied:
The origins have a `depth_type` attribute:

<img width="731" alt="screen shot 2018-01-25 at 23 08 13" src="https://user-images.githubusercontent.com/800487/35415072-a875faea-0224-11e8-8757-ed1c83f56963.png">

But I'm not sure how that would map to the values you mentioned. There is also an `origin_type` field but that does not seem to be applicable here.

<img width="710" alt="screen shot 2018-01-25 at 23 10 08" src="https://user-images.githubusercontent.com/800487/35415171-e812d524-0224-11e8-9c96-493772e383ad.png">


`loc_ind` make little sense to be to be honest. If there is a location it must have been located somehow.

calum-chamberlain replied:
Hmm, I'm not sure the `depth_ind` maps well onto those either.

I think `loc_ind` is used to ensure that the location is not re-calculated, e.g. it was calculated once and the user is happy with it, or it is a known location (like a blast) and should be fixed...

krischer replied:
I also don't really know how to map these things. I guess there just is no fully suitable mapping. The origin's also have an `epicenter_fixed` boolean field which might be suitable.

calum-chamberlain replied:
:+1: that sounds like where it might go. Excellent, thanks!

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -924,41 +1221,61 @@ def _write_nordic(event, filename, userid='OBSP', evtype='L', outdir='.',
+        str(evtime.minute).rjust(2) + ' ' + str(evtime.second).rjust(2) + '.' +
+        str(evtime.microsecond).ljust(1)[0:1] + ' ' + evtype.ljust(2) +
+        lat.rjust(7) + ' ' + lon.rjust(7) + depth.rjust(5) + agency.rjust(5) +
+        ksta.rjust(3) + timerms.rjust(4))
```
This is for one really pretty slow as each `+` generates an intermediate string object. Could this not just be done with the `.format()` method of string objects?

calum-chamberlain replied:
It definitely could and should! Will do.

krischer commented on obspy/io/nordic/core.py:
```diff
@@ -973,6 +1290,124 @@ def _write_nordic(event, filename, userid='OBSP', evtype='L', outdir='.',
+    Generate the two lines required for moment tensor solutions in Nordic.
+    """
+    # First line contains hypocenter info, second contains tensor info
+    lines = [list(' ' * 79 + 'M'), list(' ' * 79 + 'M')]
```
I've not seen this before but its actually kind of a nice way to get a mutable string :smile:

calum-chamberlain commented:
Thanks @krischer I will get on those before merging (as well as actually finished the confidence ellipsoid functions).

calum-chamberlain commented:
I *think* I might have got it working - I think I was using the wrong form of the [rotation matrix](https://en.wikipedia.org/wiki/Euler_angles#Conversion_to_other_orientation_representations) for Tait-Bryon angles.  I was using the ZYX form (which I was led to from the [QuakeML description](https://quake.ethz.ch/quakeml/QuakeML2.0/BasicEventDescriptionTypesDiscussion#class_ConfidenceEllipsoid)), but it looks like I was going the wrong-way, so changing to XYZ appears to give the correct answers for the simple cases I have tested.

megies commented on obspy/io/nordic/core.py:
```diff
@@ -77,13 +77,16 @@ def _is_sfile(sfile, encoding='latin-1'):
         try:
             head_line = tags['1'][0][0]
-        except KeyError:
+        except IndexError:
```
without knowing details.. there's three nested `get` operations in one single try/except.. looks fragile

calum-chamberlain replied:
Thanks for having a look @megies, but I'm sorry, I don't get this comment, could you explain a bit more (sorry!).

calum-chamberlain commented:
@megies and @krischer, is there anything else you would like me to change in this? From my side it looks good to go, but I haven't moved the confidence-ellipsoid functions to somewhere more general - I didn't see any functions in core.event.Event, so thought I might have been looking in the wrong place.

@megies, I didn't understand your comment on the nested `get` operations, happy to change how that is implemented though if needed?

krischer commented:
Sorry for the delay. Can be merged from my point of view. @megies what do you think?

megies commented on obspy/io/nordic/core.py:
```diff
@@ -213,94 +241,162 @@ def _nortoevmag(mag_type):
 
 
-def readheader(sfile):
+def xyz_to_confidence_ellipsoid(errors):
```
Maybe this should be moved to `core.event.base`, so that it has more visibility and other people can reuse it? It could even be a staticmethod on `ConfidenceEllipsoid` to create a new such object?

megies commented:
I didn't look at every detail but I guess it's OK.. personally, I'd move the confidence ellipsoid routine to core.event so that it can be reused more easily..

krischer commented:
I agree with moving the confidence ellipsoid routine as it might be useful to have around in a more accessible place.

calum-chamberlain commented:
Hi all.  Moved confidence ellipsoid things and added more tests, which showed that for more complicated cases it isn't working.  I have failed to fix it, and haven't been able to get help.  I suggest that I take out the confidence ellipsoid conversions and leave this with all the other little patches to nordic.
Sorry for taking up so much time.

megies commented:
The confidence ellipsoid stuff would've been nice, but yeah, whatever works for you.. you're doing the work in here.. ;-)

calum-chamberlain commented:
I've removed the confidence ellipsoid things - I will work on them outside of a PR for a while and see where I get, but in the mean-time I think the things in here should be included in obspy.  Looks like appveyor fails are not from my things. Good to go AFAIK, sans confidenceEllipsoid conversion.

krischer commented:
> I think the things in here should be included in obspy

I agree - this has been ready for a while aside the confidence ellipsoid. Let's get this merged. As always: thanks a bunch!



> Looks like appveyor fails are not from my things.

Yep - this being worked on here: https://github.com/obspy/obspy/pull/2089
